petab/core.py

```diff
@@ -453,39 +456,110 @@ def get_optimization_to_simulation_parameter_mapping(
-        _apply_overrides(
-            overrides, row.simulationConditionId,
-            row.observableId, override_type='noise')
+            mapping[condition_ix][par_sim_ix] = override
+
+    for condition_ix, condition in simulation_conditions.iterrows():
+        cur_measurement_df = get_rows_for_condition(measurement_df, condition)
+        for _, row in cur_measurement_df.iterrows():
+            # we trust that the number of overrides matches (see above)
+            overrides = split_parameter_replacement_list(
+                row.observableParameters)
+            _apply_overrides(
+                overrides, condition_ix,
+                row.observableId, override_type='observable')
+            overrides = split_parameter_replacement_list(row.noiseParameters)
+            _apply_overrides(
+                overrides, condition_ix,
+                row.observableId, override_type='noise')
+
+    handle_missing_overrides(mapping, measurement_df.observableId.unique())
 
     return mapping
 
 
+def get_simulation_conditions(measurement_df):
+    """
+    Create a table of separate simulation conditions. A simulation condition
+    is a specific combination of simulationConditionId and
+    preequilibrationConditionId.
+    """
+    # find columns to group by (i.e. if not all nans).
+    # can be improved by checking for identical condition vectors
+    grouping_cols = get_notnull_columns(
+        measurement_df,
+        ['simulationConditionId', 'preequilibrationConditionId'])
+
+    # group by cols and return dataframe containing each combination
+    # of those rows only once (and an additional counting row)
+    simulation_conditions = measurement_df.groupby(
+        grouping_cols).size().reset_index()
+
+    return simulation_conditions
+
+
+def get_rows_for_condition(measurement_df, condition):
+    """
+    Extract rows in `measurement_df` for `condition` according
+    to the grouping columns present in `condition`.
+
+    Returns
+    -------
+
+    cur_measurement_df: pd.DataFrame
+        The subselection of rows in `measurement_df` for the
+        condition `condition.
+    """
+    # filter rows for condition
+    row_filter = 1
+    # check for equality in all grouping cols
+    if 'preequilibrationConditionId' in condition:
+        row_filter = (measurement_df.preequilibrationConditionId ==
+                      condition.preequilibrationConditionId) & row_filter
+    if 'simulationConditionId' in condition:
+        row_filter = (measurement_df.simulationConditionId ==
+                      condition.simulationConditionId) & row_filter
+
+    # apply filter
+    cur_measurement_df = measurement_df.loc[row_filter, :]
+
+    return cur_measurement_df
+
+
+def handle_missing_overrides(mapping_par_opt_to_par_sim, observable_ids):
+    """
+    Find all observable parameters and noise parameters that were not mapped,
+    and set their mapping to some arbitrary numeric value.
```

Comment:
```suggestion
    and set their mapping to NaN.
```

Reply:
yes, probably better to make explicit what we do here.